### DISCUSSION TITLE
The direction of WebThrottle V2

### DISCUSSION THREAD

**Discussion by dcyoung-dev** [DCC-EX/EX-WebThrottle2#26] [Ideas]
A few decisions have been made recently regarding WebThrottle V2.

- The project will be written using [Vue](https://vuejs.org/) and [TypeScript](https://www.typescriptlang.org/).
- The outcome will be a web application aimed at new comers to the hobby, who want to get setup quickly and easily.

To make this offering possible we first need to reach feature parity with the original WebThrottle. This will be the focus of the first release.

If you have any suggestions or would like to contribute, please get in touch.

### Receive and Send messages

- [x] Works without USB connection

### Connect to Command Station

- [x] USB serial only

### Toggle power

- [x] Only simple power commands `<0>` and `<1>`

### Add a Throttle

- [ ] Add `cv` and `displayName`
- [ ] Add `speed` and `direction` controls

### Add Throttle Function Buttons

- [ ] Only the default F1 to F28 functions
- [ ] All latch switching

### Add Locomotive

- [ ] Add `cv`, `displayName`
- [ ] Populate Throttles from Locomotives

### Customise Locomotive Function Buttons

- [ ] Allow each function button to be configured as a `momentary` or `latch` switch
- [ ] Allow each function button to be renamed (`displayName`)

**Comment by Asbelos** [comment]
Not sure how you mean  "send messages" .

We have a Poc that the cs will support wrbsockets. Please make sure connection can be usb-serial or websocket.  ( its just a matter of which comms function is called, the commands are identical. )

Do we really need ways of editing and saving functions when we have control of that from the cs rosters. 

Use of roster, turnout and routes is not discussed. 

Drive away function?

**Reply by dcyoung-dev** [reply]
Keeping it simple and focused for now 🤞

By send messages I mean something really simple see #25 as an example 👌

We're going to avoid websockets for just now - we should build the USB connection in such away that it can be swapped out for something like websockets in the future. 

I'd love to go straight for populating from the Roster (it's what I've done in my version) but as we're trying to deprecate the original WebThrottle, we want to emulate the current features. 
Also, for beginners it'll be easier for them to do this through a web UI rather than a code editor to make the Roster 👍

Turnouts and routes are not currently available in WebThrottle. Turnouts would be next on my list, with routes feeling a bit more advanced - needs some external code editing. 

All of your suggestions are spot on for where we want to get to eventually and for now we want to focus on a few fundamentals 🤞

**Comment by Kcsmith0708** [comment]
Hi Dave,
Thanks for heading this up.

The Web-Throttle at a minimum should have
A) have the Power button GREEN when On, and RED when Off.
B) a serial command line that has the ability to enter commands without a < > brackets.  
C) ability for the Send Command line to scroll up & down and re-display a previously entered command so we don't have to retype it every time.  Similar to what the Arduino IDE serial monitor does now.
D) have the Whistle/Horn F2 button Non-Latching.

Future possible features;
+Add separate user defined EXRAIL push buttons displayed on the throttle. 
To easily start Automation or Set Routes and operate Accessories. 
Just need to be able to type a short label "name" and point to and load a </START n [n]> command
Using the EXRAIL Automation or Route "description" might be a little too long for a initial throttle label.

bug; currently Serial Response to the script output are being cut short and not fully displaying.
+ Allow for EXRAIL PRINT function to  fully display and
     be able to scroll the serial responses backward or forward to see previous responses displayed.
     
Thx,
Kevin

**Comment by Asbelos** [comment]
If the webthrottle is to have methods for creating your own roster/functions  then please make sure that it can  list them in a way that can be pasted into myAuromation.h later.

**Reply by dcyoung-dev** [reply]
I like this idea - it could work as some sort of exporter 👌
It's not a concern for right now 👌
